tests/agent_server/deploy/e2e/test_autostarted.py

```diff
@@ -244,3 +252,677 @@ def repaired() -> bool:
+    :param ensure_resource_tracker_is_started: The fixture that creates a Resource Tracker process
+    """
+
+    def is_consistent_state(current_processes: list[psutil.Process], process_mapping: dict[str, list[psutil.Process]]) -> bool:
+        """
+        Returns True if we are in one of the above situation, False otherwise
+
+        :param current_processes: Current processes
+        :param process_mapping: Current mapping of processes
+        """
+        match len(current_processes):
+            case 2:
+                return len(process_mapping["pg_ctl"]) == 1 and len(process_mapping["python"]) == 1
+            case 3:
+                return (
+                    len(process_mapping["pg_ctl"]) == 1
+                    and len(process_mapping["python"]) == 2
+                    or len(process_mapping["pg_ctl"]) == 1
+                    and len(process_mapping["python"]) == 1
+                    and len(process_mapping["inmanta fork server"]) == 1
+                )
+            case 4:
+                return (
+                    len(process_mapping["pg_ctl"]) == 1
+                    and len(process_mapping["python"]) == 2
+                    and len(process_mapping["inmanta fork server"]) == 1
+                )
+            case _:
+                return False
+
+    async def wait_for_consistent_state() -> bool:
+        """
+        Wait for a consistent state
+        """
+        pre_start_children = get_process_state(current_pid)
+        assert len(pre_start_children) == 1
+        assert len(pre_start_children.values()) == 1
+        old_children = list(pre_start_children.values())[0]
+        mapping_process = retrieve_mapping_process(old_children)
+
+        return is_consistent_state(current_processes=old_children, process_mapping=mapping_process)
+
+    current_pid = os.getpid()
+    await agent.stop()
+    await retry_limited(wait_for_consistent_state, timeout=10)
+    return current_pid
+
+
+def get_process_state(current_pid: int) -> dict[psutil.Process, list[psutil.Process]]:
+    """
+    Retrieves the current list of processes running under this process
+
+    :param current_pid: The PID of this process
+    """
+    return {process: process.children(recursive=True) for process in psutil.process_iter() if process.pid == current_pid}
+
+
+@pytest.mark.parametrize(
+    "auto_start_agent,should_time_out,time_to_sleep,", [(True, False, 2), (True, True, 120)]
+)  # this overrides a fixture to allow the agent to fork!
+async def test_halt_deploy(
+    snippetcompiler,
+    server,
+    ensure_consistent_starting_point: int,
+    client,
+    clienthelper,
+    environment,
+    no_agent_backoff,
+    auto_start_agent: bool,
+    should_time_out: bool,
+    time_to_sleep: int,
+):
+    """
+    Verify that the new scheduler can actually halt an ongoing deployment and can resume it when the user requests it
+    """
+    current_pid = ensure_consistent_starting_point
```

Comment:
this could do with some inline comments to show what is going on, for all of them